lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -200,10 +202,50 @@ getTransaction = isolate 43 $ do
             other -> fail $ "Invalid witness type: " ++ show other
 
 
+putTransaction :: (Tx, [TxWitness]) -> Put
+putTransaction (tx, witnesses) = do
+    putTokenTransfer tx
+    mapM_ putWitness witnesses
+
+putWitness :: TxWitness -> Put
+putWitness witness =
+    case witness of
+        PublicKeyWitness xPub (Hash sig) -> do
+            putWord8 1
+            putByteString xPub
+            putByteString sig
+        -- TODO: note that we are missing new address type witness:
+        --  * https://github.com/input-output-hk/rust-cardano/blob/3524cfe138a10773caa6f0effacf69e792f915df/chain-impl-mockchain/doc/format.md#witnesses
+        --  * https://github.com/input-output-hk/rust-cardano/blob/e0616f13bebd6b908320bddb1c1502dea0d3305a/chain-impl-mockchain/src/transaction/witness.rs#L23
+        ScriptWitness _ -> error "unimplemented: serialize script witness"
+        RedeemWitness _ -> error "unimplemented: serialize redeem witness"
+
+
 {-------------------------------------------------------------------------------
                             Common Structure
 -------------------------------------------------------------------------------}
 
+putTokenTransfer :: Tx -> Put
+putTokenTransfer (Tx inputs outputs) = do
+    putWord8 $ fromIntegral $ length inputs
+    putWord8 $ fromIntegral $ length outputs
+    mapM_ putInput inputs
+    mapM_ putOutput outputs
+  where
+    putInput (TxIn inputId inputIx) = do
+        -- NOTE: special value 0xff indicates account spending
+        -- only old utxo/address scheme supported for now
+        putWord8 $ fromIntegral inputIx
+        putByteString $ getHash inputId
+    putOutput (TxOut address coin) = do
+        putAddress address
+        putWord64be $ getCoin coin
+    putAddress address = do
+        -- NOTE: only single address supported for now
+        putWord8 single
+        -- TODO(akegalj) this should be `encodeAddress (Proxy Jormungandr) address`
+        putByteString "addressBytestring" -- address
```

Comment:
I think this line should look like 
```haskell
putByteString $ encodeAddress (Proxy Jormungandr) address
```

but can't import `Jormungandr.Compatibility` to get `Jormungandr` because of the cyclic imports. I can move it to another module or see other ways to rearange modules/imports. Alternativelly we can pass the type variable via `Proxy` if we define method as

```haskell
putTokenTransfer :: Proxy x -> Tx -> Put
```

but this doesn't make sense to me as its well known there is only one type that should go here.

Reply:
> encodeAddress

🤔 The addresses are just raw ByteStrings though, right?

Reply:
aha - they might not be encoded as I thought - will check

Reply:
shouldn't be relevant as address should be bytestring